### PR TITLE
Initial, compiling, untested port of typed abts from SML based on locally-nameless terms

### PR DESCRIPTION
**Pull request by wemrysi**
This is a compiling, but untested port of [jonsterling/sml-typed-abts](https://github.com/jonsterling/sml-typed-abts). It isn't ready for production yet, but this PR is intended to get what we have so far merged upstream for the purposes of integration.

Next steps (not an exhaustive list)
- Tests
- Examples
- Performance
- Ensure that `AbtError` doesn't include implementation-specific (i.e. `LN`) cases, generalize if so
- [Monocle](https://github.com/julien-truffaut/Monocle) lenses for datatypes

### CODE REVIEW

File: .travis.yml [hunk 1]
```diff
@@ -0,0 +1,6 @@
+language: scala
+scala:
+  - 2.11.7
+jdk:
+  - oraclejdk7
```
Comment (sellout) [inline .travis.yml:5]: We don’t mention the JDK in the quasar repo … should we be?

File: build.sbt [hunk 2]
```diff
@@ -39,7 +39,7 @@ libraryDependencies ++= Seq(
 )
 
 // Wartremover
-wartremoverErrors ++= Warts.unsafe
+wartremoverErrors ++= Warts.unsafe.filterNot(_.clazz == Wart.Throw.clazz)
```
Comment (sellout) [inline build.sbt]: Can we use `Warts.allBut(Wart.Throw)`? Also, where are we using `throw`?

Comment (wemrysi) [inline build.sbt]: We can, though `Warts.unsafe` are described as the "stable" checks (i.e. less false-positives). `throw` is triggered by pattern matching.

Comment (sellout) [inline build.sbt]: I’d rather try to use them all, then be explicit about disabling the ones that cause us trouble.

File: .travis.yml (hunk 1 above)
Comment (wemrysi) [inline .travis.yml:5]: I'd say we should, otherwise we're running against whatever the travis default jdk is (in this case it happens to be Oracle JDK 7).

**Pull request merged**